### PR TITLE
PASS: Logic and Validation add methods in result unit tests

### PR DESCRIPTION
**Pull request by Ravshan9812**

### CODE REVIEW

File: EssayAnalyzer.Api/Services/Foundation/Results/ResultService.Exception.cs [hunk 1]
```diff
@@ -1,6 +1,38 @@
...
+        catch(InvalidResultException invalidResultException)
+        {
+            throw CreateAndLogValidationException(invalidResultException);
+        }
+    }
+
+    private Exception CreateAndLogValidationException(Xeption exception)
+    {
+        var resultValidationException =
+                    new ResultValidationException(exception);
```
Comment (azam-amonov) [inline EssayAnalyzer.Api/Services/Foundation/Results/ResultService.Exception.cs]: Make single line code on 
var resultValidationException =
                    new ResultValidationException(exception);

File: EssayAnalyzer.Api/Services/Foundation/Results/ResultService.Validaton.cs [hunk 2]
```diff
@@ -1,6 +1,54 @@
...
 public partial class ResultService
 {
-    
+    private static void ValidateResultOnAdd(Result result)
+    {
+        ValidateResultIsNotNull(result);
+
+        Validate(
+                (Rule: IsInvalid(result.Id), Parametr: nameof(Result.Id)),
+                (Rule: IsInvalid(result.Feedback), Parametr: nameof(Result.Feedback)));
```
Comment (azam-amonov) [inline EssayAnalyzer.Api/Services/Foundation/Results/ResultService.Validaton.cs:16]: Add Validation Rule for EssayId

File: EssayAnalyzer.Api/Services/Foundation/Results/ResultService.Validaton.cs [hunk 3]
```diff
@@ -1,6 +1,54 @@
...
+        if (result is null)
+        {
+            throw new NullResultException();
+        }
+    }
+
+    private static dynamic IsInvalid(Guid id) => new
+    {
+        Condition = id == Guid.Empty,
+        Message = "Id is required."
```
Comment (azam-amonov) [inline EssayAnalyzer.Api/Services/Foundation/Results/ResultService.Validaton.cs]: Remove dot (.) from message

File: EssayAnalyzer.Api/Services/Foundation/Results/ResultService.Validaton.cs [hunk 4]
```diff
@@ -1,6 +1,54 @@
...
+    };
+
+    private static dynamic IsInvalid(string text) => new
+    {
+        Condition = String.IsNullOrWhiteSpace(text),
+        Message = "Text is required."
+    };
```
Comment (azam-amonov) [inline EssayAnalyzer.Api/Services/Foundation/Results/ResultService.Validaton.cs:37]: Remove dot (.)  message

**Review by jaysindarov — COMMENTED** [review]
image
use just [Theory] instead of [Xunit.Theory]

File: EssayAnalyzer.Api/Services/Foundation/Results/ResultService.Validaton.cs [hunk 5]
```diff
@@ -1,6 +1,54 @@
...
+        foreach ((dynamic rule, string parametr) in validations)
+        {
+            if (rule.Condition)
+            {
+                invalidResultException.UpsertDataList(
+                    key: parametr,
+                    value: rule.Message);
+            }
+        }
+
```
Comment (azam-amonov) [inline EssayAnalyzer.Api/Services/Foundation/Results/ResultService.Validaton.cs:52]: Remove empty line

File: EssayAnalyzer.UnitTest/Services/Foundation/Results/ResultServiceTests.Logic.Add.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace EssayAnalyzer.UnitTest.Services.Foundation.Results
+{
+    public partial class ResultServiceTests
```
Comment (azam-amonov) [inline EssayAnalyzer.UnitTest/Services/Foundation/Results/ResultServiceTests.Logic.Add.cs]: Remove unused libraries, clean namespaces

File: EssayAnalyzer.UnitTest/Services/Foundation/Results/ResultServiceTests.Logic.Add.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        public async Task ShouldAddResultAsync()
+        {
+            // given
+            Result randomResult = CreateRandomResult();
+            Result inputResult = randomResult;
+            Result persistedResult = inputResult;
+            Result expectedResult = persistedResult.DeepClone();
+
+            this.storageBrokerMock.Setup(broker =>
+            broker.InsertResultAsync(inputResult))
```
Comment (azam-amonov) [inline EssayAnalyzer.UnitTest/Services/Foundation/Results/ResultServiceTests.Logic.Add.cs:21]: Add Tab for second line

File: EssayAnalyzer.UnitTest/Services/Foundation/Results/ResultServiceTests.Validation.Add.cs [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace EssayAnalyzer.UnitTest.Services.Foundation.Results
+{
+    public partial class ResultServiceTests
```
Comment (azam-amonov) [inline EssayAnalyzer.UnitTest/Services/Foundation/Results/ResultServiceTests.Validation.Add.cs]: Make clean namespaces here

File: EssayAnalyzer.UnitTest/Services/Foundation/Results/ResultServiceTests.Validation.Add.cs [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+            var invalidResultException = new InvalidResultException();
+
+            invalidResultException.AddData(
+                key: nameof(Result.Id),
+                values: "Id is required.");
+
+            invalidResultException.AddData(
+                key: nameof(Result.Feedback),
+                values: "Text is required.");
+
```
Comment (azam-amonov) [inline EssayAnalyzer.UnitTest/Services/Foundation/Results/ResultServiceTests.Validation.Add.cs:63]: Add 
 invalidResultException.AddData(
                key: nameof(Result.EssayId),
                values: "Id is required.");

File: EssayAnalyzer.UnitTest/Services/Foundation/Results/ResultServiceTests.cs [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Text;
+using System.Threading.Tasks;
+using Tynamix.ObjectFiller;
+using Xeptions;
+
+namespace EssayAnalyzer.UnitTest.Services.Foundation.Results
+{
```
Comment (azam-amonov) [inline EssayAnalyzer.UnitTest/Services/Foundation/Results/ResultServiceTests.cs:18]: Clean Namespaces

**Review by azam-amonov — CHANGES_REQUESTED** [review]
### First of all!
- Before commit PASSED Test commit FAIL

**Review by azam-amonov — APPROVED** [review]

**Review by jaysindarov — COMMENTED** [review]
image

Don't use just Exception as type of your method because it's not a service exception it is validation exception

**Review by jaysindarov — COMMENTED** [review]
image

remove empty space in 22nd line

**Review by jaysindarov — APPROVED** [review]

**Pull request merged**